File: labs/OP/lab_03/library/memoization.py
from functools import wraps
from time import monotonic
from typing import Any, Callable


def memoize(
    *,
    max_size: int | None = None,
    policy: str = "lru",
    ttl_seconds: float | None = None,
    custom_evictor: Callable[[dict[Any, dict[str, Any]]], Any | None] | None = None,
):
    if max_size is not None and max_size < 1:
        raise ValueError("max_size must be None or a positive integer")

    if policy not in {"lru", "lfu", "ttl", "custom"}:
        raise ValueError("policy must be one of: lru, lfu, ttl, custom")

    if ttl_seconds is not None and ttl_seconds <= 0:
        raise ValueError("ttl_seconds must be positive")

    if policy == "ttl" and ttl_seconds is None:
        raise ValueError("ttl_seconds is required when policy='ttl'")

    if policy == "custom" and custom_evictor is None:
        raise ValueError("custom_evictor is required when policy='custom'")

    cache: dict[Any, dict[str, Any]] = {}

    def _make_key(args: tuple[Any, ...], kwargs: dict[str, Any]):
        return (args, tuple(sorted(kwargs.items())))

    def _is_expired(entry: dict[str, Any], now: float) -> bool:
        if policy != "ttl" or ttl_seconds is None:
            return False
        return now - entry["created_at"] >= ttl_seconds

    def _prune_expired(now: float) -> None:
        if policy != "ttl" or ttl_seconds is None:
            return
        expired_keys = [key for key, entry in cache.items() if _is_expired(entry, now)]
        for key in expired_keys:
            cache.pop(key, None)

    def _evict_one() -> None:
        if not cache:
            return

        if policy == "lru":
            key_to_remove = min(cache, key=lambda key: cache[key]["last_access"])
            cache.pop(key_to_remove, None)
            return

        if policy == "lfu":
            key_to_remove = min(
                cache,
                key=lambda key: (cache[key]["access_count"], cache[key]["last_access"]),
            )
            cache.pop(key_to_remove, None)
            return

        if policy == "ttl":
            key_to_remove = min(cache, key=lambda key: cache[key]["created_at"])
            cache.pop(key_to_remove, None)
            return

        eviction_result = custom_evictor(cache) if custom_evictor is not None else None
        if eviction_result is None:
            key_to_remove = min(cache, key=lambda key: cache[key]["last_access"])
            cache.pop(key_to_remove, None)
            return

        if eviction_result in cache:
            cache.pop(eviction_result, None)
            return

        key_to_remove = min(cache, key=lambda key: cache[key]["last_access"])
        cache.pop(key_to_remove, None)

    def decorator(func: Callable[..., Any]):
        @wraps(func)
        def wrapper(*args, **kwargs):
            now = monotonic()
            _prune_expired(now)

            key = _make_key(args, kwargs)
            if key in cache:
                entry = cache[key]
                if _is_expired(entry, now):
                    cache.pop(key, None)
                else:
                    entry["last_access"] = now
                    entry["access_count"] += 1
                    return entry["value"]

            result = func(*args, **kwargs)

            while max_size is not None and len(cache) >= max_size:
                _evict_one()

            cache[key] = {
                "value": result,
                "created_at": now,
                "last_access": now,
                "access_count": 1,
            }
            return result

        return wrapper

    return decorator

File: courseworks/DecisionDex/core/memoization.js
var Memoization = {
  memoize: function (fn, options) {
    options = options || {};
    var maxSize = options.maxSize == null ? null : options.maxSize;
    var policy = (options.policy || options.strategy || "lru").toLowerCase();
    var ttlSeconds = options.ttlSeconds == null ? options.ttl : options.ttlSeconds;
    var customEvictor = options.customEvictor || null;

    if (maxSize !== null && maxSize < 1) {
      throw new Error("maxSize must be null or a positive integer");
    }
    if (["lru", "lfu", "ttl", "custom"].indexOf(policy) === -1) {
      throw new Error("policy must be one of: lru, lfu, ttl, custom");
    }
    if (ttlSeconds != null && ttlSeconds <= 0) {
      throw new Error("ttlSeconds must be positive");
    }
    if (policy === "ttl" && ttlSeconds == null) {
      throw new Error("ttlSeconds is required when policy='ttl'");
    }
    if (policy === "custom" && !customEvictor) {
      throw new Error("customEvictor is required when policy='custom'");
    }

    var cache = new Map();

    function makeKey(args, kwargs) {
      return JSON.stringify([args, kwargs]);
    }

    function parseKey(key) {
      try {
        var parsed = JSON.parse(key);
        return { args: parsed[0] || [], kwargs: parsed[1] || {} };
      } catch (e) {
        return { args: [], kwargs: {} };
      }
    }

    function isExpired(entry, now) {
      if (policy !== "ttl" || ttlSeconds == null) {
        return false;
      }
      return now - entry.createdAt >= ttlSeconds * 1000;
    }

    function pruneExpired(now) {
      if (policy !== "ttl" || ttlSeconds == null) {
        return;
      }
      var toDelete = [];
      cache.forEach(function (entry, key) {
        if (isExpired(entry, now)) {
          toDelete.push(key);
        }
      });
      toDelete.forEach(function (key) {
        cache.delete(key);
      });
    }

    function evictOne() {
      if (cache.size === 0) {
        return;
      }

      var keyToRemove = null;

      if (policy === "lru") {
        cache.forEach(function (entry, key) {
          if (!keyToRemove || entry.lastAccess < cache.get(keyToRemove).lastAccess) {
            keyToRemove = key;
          }
        });
      } else if (policy === "lfu") {
        cache.forEach(function (entry, key) {
          if (!keyToRemove) {
            keyToRemove = key;
            return;
          }
          var current = cache.get(keyToRemove);
          if (
            entry.accessCount < current.accessCount ||
            (entry.accessCount === current.accessCount && entry.lastAccess < current.lastAccess)
          ) {
            keyToRemove = key;
          }
        });
      } else if (policy === "ttl") {
        cache.forEach(function (entry, key) {
          if (!keyToRemove || entry.createdAt < cache.get(keyToRemove).createdAt) {
            keyToRemove = key;
          }
        });
      } else {
        var snapshot = {};
        cache.forEach(function (entry, key) {
          snapshot[key] = {
            value: entry.value,
            createdAt: entry.createdAt,
            lastAccess: entry.lastAccess,
            accessCount: entry.accessCount,
            keyData: parseKey(key),
          };
        });
        var customKey = customEvictor ? customEvictor(snapshot) : null;
        if (customKey && cache.has(customKey)) {
          keyToRemove = customKey;
        }
      }

      if (!keyToRemove) {
        cache.forEach(function (_entry, key) {
          if (!keyToRemove) {
            keyToRemove = key;
          }
        });
      }

      if (keyToRemove) {
        cache.delete(keyToRemove);
      }
    }

    return function () {
      var args = Array.prototype.slice.call(arguments);
      var kwargs = {};
      var now = Date.now();

      pruneExpired(now);

      var key = makeKey(args, kwargs);
      if (cache.has(key)) {
        var entry = cache.get(key);
        if (!isExpired(entry, now)) {
          entry.lastAccess = now;
          entry.accessCount += 1;
          return entry.value;
        }
        cache.delete(key);
      }

      var result = fn.apply(this, args);

      while (maxSize !== null && cache.size >= maxSize) {
        evictOne();
      }

      cache.set(key, {
        value: result,
        createdAt: now,
        lastAccess: now,
        accessCount: 1,
      });
      return result;
    };
  },
};
